File: packages/ui-flutter/lib/src/code_block.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'theme.dart';

class KinetixCodeBlockFile {
  const KinetixCodeBlockFile({required this.name, required this.code});

  final String name;
  final String code;
}

/// Mirrors `packages/ui/src/components/code-block.tsx`: a monospaced code
/// display with a copy button and, for more than one file, a tab strip.
/// Presentational — no syntax highlighting, same as the other ports. Copy
/// uses Flutter's `Clipboard` (cross-platform).
class KinetixCodeBlock extends StatefulWidget {
  KinetixCodeBlock({super.key, required String code, String filename = '', this.hideCopy = false})
      : files = [KinetixCodeBlockFile(name: filename, code: code)];

  const KinetixCodeBlock.files({super.key, required this.files, this.hideCopy = false});

  final List<KinetixCodeBlockFile> files;
  final bool hideCopy;

  @override
  State<KinetixCodeBlock> createState() => _KinetixCodeBlockState();
}

class _KinetixCodeBlockState extends State<KinetixCodeBlock> {
  int _active = 0;
  bool _copied = false;

  KinetixCodeBlockFile get _current =>
      widget.files.isEmpty ? const KinetixCodeBlockFile(name: '', code: '') : widget.files[_active.clamp(0, widget.files.length - 1)];

  bool get _hasHeader => widget.files.length > 1 || _current.name.isNotEmpty;

  Future<void> _copy() async {
    await Clipboard.setData(ClipboardData(text: _current.code));
    if (!mounted) return;
    setState(() => _copied = true);
    await Future<void>.delayed(const Duration(milliseconds: 1500));
    if (mounted) setState(() => _copied = false);
  }

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);

    final Widget copyButton = GestureDetector(
      onTap: _copy,
      behavior: HitTestBehavior.opaque,
      child: Padding(
        padding: const EdgeInsets.all(6),
        child: Icon(_copied ? Icons.check : Icons.copy, size: 14, color: c.mutedForeground),
      ),
    );

    return ClipRRect(
      borderRadius: BorderRadius.circular(8),
      child: DecoratedBox(
        decoration: BoxDecoration(
          color: c.muted,
          border: Border.all(color: c.input, width: 1),
          borderRadius: BorderRadius.circular(8),
        ),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            if (_hasHeader)
              DecoratedBox(
                decoration: BoxDecoration(
                  color: c.background,
                  border: Border(bottom: BorderSide(color: c.input, width: 1)),
                ),
                child: Row(
                  children: [
                    for (var i = 0; i < widget.files.length; i++)
                      GestureDetector(
                        onTap: () => setState(() => _active = i),
                        behavior: HitTestBehavior.opaque,
                        child: Container(
                          padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 8),
                          decoration: BoxDecoration(
                            border: Border(
                              bottom: BorderSide(
                                color: i == _active ? c.primary : const Color(0x00000000),
                                width: 2,
                              ),
                            ),
                          ),
                          child: Text(
                            widget.files[i].name.isEmpty ? 'code' : widget.files[i].name,
                            style: TextStyle(
                              fontSize: 13,
                              color: i == _active ? c.foreground : c.mutedForeground,
                            ),
                          ),
                        ),
                      ),
                    const Spacer(),
                    if (!widget.hideCopy) copyButton,
                  ],
                ),
              ),
            Stack(
              children: [
                SingleChildScrollView(
                  scrollDirection: Axis.horizontal,
                  child: Padding(
                    padding: const EdgeInsets.all(12),
                    child: SelectableText(
                      _current.code,
                      style: TextStyle(fontSize: 13, fontFamily: 'monospace', color: c.foreground),
                    ),
                  ),
                ),
                if (!_hasHeader && !widget.hideCopy)
                  Positioned(right: 6, top: 6, child: copyButton),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/CodeBlock.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * KinetixCodeBlock — mirrors `packages/ui/src/components/code-block.tsx`:
 * a code display with a copy button and, for more than one file, a tab
 * strip. Presentational only, same as the source — bring your own syntax
 * highlighting by passing pre-highlighted text if needed; this just
 * renders `code`/`files[].code` as plain monospace text. Copy uses
 * Compose's own `LocalClipboardManager` — no new dependency, same "the
 * platform already has this" reuse as `KinetixToaster` wrapping
 * `SnackbarHostState`. Horizontal overflow wraps [KinetixScrollArea],
 * genuine reuse rather than a new scroll implementation. The
 * copy-button glyph is the plain text "Copy"/"Copied" swap, same
 * convention `KinetixPasswordInput`'s show/hide toggle uses.
 */
data class KinetixCodeBlockFile(val name: String, val code: String)

@Composable
fun KinetixCodeBlock(
    code: String,
    modifier: Modifier = Modifier,
    filename: String? = null,
    files: List<KinetixCodeBlockFile> = emptyList(),
    hideCopy: Boolean = false,
) {
    val colors = KinetixColorScheme.current
    val tabs = files.ifEmpty { listOf(KinetixCodeBlockFile(name = filename.orEmpty(), code = code)) }
    var active by remember { mutableIntStateOf(0) }
    var copied by remember { mutableStateOf(false) }
    val current = tabs.getOrElse(active) { tabs[0] }
    val hasHeader = tabs.size > 1 || current.name.isNotEmpty()
    val clipboard = LocalClipboardManager.current
    val scope = rememberCoroutineScope()
    val borderWidthPx = with(LocalDensity.current) { dimensionResource(R.dimen.border_width_default).toPx() }
    val shape = RoundedCornerShape(dimensionResource(R.dimen.radius_md))

    fun copy() {
        clipboard.setText(AnnotatedString(current.code))
        copied = true
        scope.launch {
            delay(1500)
            copied = false
        }
    }

    Column(
        modifier = modifier
            .clip(shape)
            .background(colors.muted, shape)
            .border(dimensionResource(R.dimen.border_width_default), colors.border, shape),
    ) {
        if (hasHeader) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .background(colors.background)
                    .drawBehind {
                        drawLine(
                            color = colors.border,
                            start = Offset(0f, size.height),
                            end = Offset(size.width, size.height),
                            strokeWidth = borderWidthPx,
                        )
                    }
                    .padding(horizontal = dimensionResource(R.dimen.spacing_2)),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Row {
                    tabs.forEachIndexed { i, tab ->
                        val selected = i == active
                        Text(
                            text = tab.name.ifEmpty { "code" },
                            color = if (selected) colors.foreground else colors.mutedForeground,
                            fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp,
                            modifier = Modifier
                                .clickable { active = i }
                                .padding(
                                    horizontal = dimensionResource(R.dimen.spacing_3),
                                    vertical = dimensionResource(R.dimen.spacing_2),
                                ),
                        )
                    }
                }
                if (!hideCopy) {
                    Text(
                        text = if (copied) "Copied" else "Copy",
                        color = colors.mutedForeground,
                        fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp,
                        modifier = Modifier.clickable(onClick = ::copy),
                    )
                }
            }
        }
        Box {
            KinetixScrollArea(orientation = KinetixScrollAreaOrientation.Horizontal) {
                Text(
                    text = current.code,
                    modifier = Modifier.padding(dimensionResource(R.dimen.spacing_4)),
                    color = colors.foreground,
                    fontFamily = FontFamily.Monospace,
                    fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp,
                )
            }
            if (!hasHeader && !hideCopy) {
                Text(
                    text = if (copied) "Copied" else "Copy",
                    color = colors.mutedForeground,
                    fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp,
                    modifier = Modifier
                        .align(Alignment.TopEnd)
                        .padding(dimensionResource(R.dimen.spacing_2))
                        .clickable(onClick = ::copy),
                )
            }
        }
    }
}
